src/classes/RD2_DataMigration_BATCH.cls

```diff
@@ -94,60 +114,115 @@ public class RD2_DataMigration_BATCH implements Database.Batchable<SObject>, Sch
-            .withRelatedOpportunities(opps);
-
-        if (mapper.isConverted()) {	
-            return;	
+    private void setBatchJobId(Database.BatchableContext context) {
+        if (context == null || batchJobId != null) {
+            return;
         }
 
-        convertedRDs.add(mapper.convertToEnhancedRD());
+        batchJobId = context.getJobId();
+    }
+
+    /**
+    * @description Constructs log context
+    * @return String Log context containing batch job Created Date
+    */
+    private String constructLogContext() {
+        UTIL_BatchJobService.BatchJob batchJob = new UTIL_BatchJobService().getBatchJob(batchJobId);
 
-        // TODO Opportunity Updates
+        String formattedDateSuffix = batchJob == null 
+            ? ''
+            : ' ' + String.valueOf(batchJob.CreatedDate);
 
-        // TODO Retrieve any validation errors that blocked the conversion
+        return System.Label.RD2_DataMigrationLogContext + formattedDateSuffix;
+    }
+
+    /**
+    * @description Constructs log records
+    * @param e Exception
+    * @param rd Recurring Donation
+    * @return void
+    */
+    private void logException(Exception e, npe03__Recurring_Donation__c rd) {
+        Error__c logEntry = new Error__c(
+            Context_Type__c = logContext,
+            Datetime__c = System.now(),
+            Related_Record_ID__c = rd.Id,
+            Record_URL__c = ERR_Handler.getRecordURL(rd.Id),
+            Error_Type__c = System.Label.errErrorTypeValidationFailed,
+            Full_Message__c = e.getMessage()
+        );
+
+        ERR_Handler.processErrors(new List<Error__c>{ logEntry }, logContext);
     }
 
     /**
     * @description When all RD's have been processed, handle any required DML and error/exception logging
-    * @param sp SavePoint if a rollback is required for dry run mode
+    * @param convertedRDs Converted Recurring Donations
+    * @return void
     */
-    private void handleDmlAndLogging(Savepoint sp) {
-        // TODO Rewrite this to capture and log dml exceptions in a new logging service
-        //Database.update(convertedRDs, false);
-        update convertedRDs;
+    private void handleDmlAndLogging(List<npe03__Recurring_Donation__c> convertedRDs) {
+        if (convertedRDs.isEmpty()) {
+            return;
+        }
+
+        Savepoint sp = Database.setSavepoint();
+
+        List<Database.SaveResult> saveResults = Database.update(convertedRDs, false);
 
         if (this.isDryRunMode) {
             Database.rollback(sp);
         }
+        
+        ERR_Handler.Errors errorResult = ERR_Handler.getErrors(saveResults, convertedRDs);
+        if (errorResult.errorsExist) {            
+            for (Error__c error : errorResult.errorRecords) {
+                error.Context_Type__c = logContext;
+            }
 
-        // TODO Commit any logging records after the rollback
+            ERR_Handler.processErrors(errorResult.errorRecords, logContext);
```

Comment:
Good point, thank you!
The [Opportunity Service](https://github.com/SalesforceFoundation/Cumulus/blob/feature/RD2/src/classes/RD2_OpportunityService.cls#L94) is using the same error processing method. Does the service need to be changed as well?

Reply:
Good question. It's all about whether there is a page context at the time of the exception. For normal operations by that service, which is called asynchronously by a trigger or even the nightly batch job, this won't be an issue because there is no page context. 

I believe I mentioned to you that I'd like to _eventually_ move away from DML in that service, but that's likely not happening immediately unless we feel it's a must for GA - which I'm not sure I do. So, for the moment I think this is fine. It won't generate errors in the UI for customers given its async context.

Reply:
Have you tested this to see what the impact on the UI is when multiple errors are generated?

Reply:
Good catch! I forgot to make the change.

The errors are not displayed, however, to be on the safe side, a new method `ERR_Handler.processErrorsOnly()` has been added to process errors without displaying them on the page.